src/components/hooks/useStorage.js

```diff
@@ -0,0 +1,37 @@
+
+import { useState, useEffect } from "react";
+import { storage } from "../firebase/config"
+import { ref, uploadBytesResumable, getDownloadURL } from "firebase/storage"
+
+const useStorage = (file) => { 
+    
+    const [ progress, setProgress] = useState(0);
+    const [ error, setError ] = useState(null);
+    const [ url, setUrl ] = useState(null)
+
+
+    useEffect(() => {
+
+        const storageRef = ref(storage, `/archive/${file.name}`)
+        
+        const uploadTask = uploadBytesResumable(storageRef, file);
+        
+        uploadTask.on('state_changed', (snap) => { 
+            let percentage = (snap.bytesTransferred / snap.totalBytes) * 100;
+            setProgress(percentage)
+        }, (err) => { 
+            setError(err);
+        }, async () => { 
+            getDownloadURL(ref(storage, `/archive/${file.name}`)).then((url) => {
+
+        setUrl(url)
+  })        
+        })
+    }, [file])
+
```

Comment:
As a rule of thumb, try to avoid useEffects at all cost, especially for handling events. 
The main use of useEffect is to synchronise React state with external effects... but everything you need here is internal.

I would definitely keep the progress, url and error, those are great, and the upload logic as well. Would probably null the error first, but other than that, great.

What I would do different: drop the effect, and do an `onUpload` handler instead exported, which does the upload, straight on file change. That can be then used as an onChange handler on the actual input component.  You get to keep the state updates, but now your function is only running on even, not whenever React decides to re-run effects :)

The reason why you upload twice, is that React in strict mode re-renders twice in dev environment, to discourage you using useEffect this way :)